### ISSUE TITLE
`samba` cross-compilation fails

### ISSUE THREAD

**Issue by Majiir** [NixOS/nixpkgs#200638]
### Steps To Reproduce

On latest `master`, which was c1c640f123526630f6cecc0dd6671270e644cbc1 at time of writing:

```
nix build nixpkgs#pkgsCross.armv7l-hf-multiplatform.samba
nix build nixpkgs#pkgsCross.aarch64-multiplatform.samba
```

### Build log

```
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/awaqq5xswhgsi3nwgasvp56g9sas96y6-samba-4.17.2.tar.gz
source root is samba-4.17.2
setting SOURCE_DATE_EPOCH to timestamp 1666608797 of file samba-4.17.2/docs/manpages/winexe.1
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
applying patch /nix/store/ifgilzjj9kyv2m1kik1rr1i34i8p9h1d-4.x-no-persistent-install.patch
patching file ctdb/wscript
Hunk #1 succeeded at 842 (offset 28 lines).
Hunk #2 succeeded at 880 (offset 28 lines).
applying patch /nix/store/npsh8b7fc3hvb3vag41fmbxsq7m1hjxy-patch-source3__libads__kerberos_keytab.c.patch
patching file source3/libads/kerberos_keytab.c
Hunk #2 succeeded at 332 (offset 249 lines).
applying patch /nix/store/w16z1hx9kzlz7v42f4x1wy0cdsnlz5np-4.x-no-persistent-install-dynconfig.patch
patching file dynconfig/wscript
Hunk #1 succeeded at 435 with fuzz 1 (offset 16 lines).
applying patch /nix/store/xlss48c1b1z26qjm76qg7dqk6lm58zvi-4.x-fix-makeflags-parsing.patch
patching file buildtools/wafsamba/samba_utils.py
patching script interpreter paths in ./buildtools/bin
./buildtools/bin/waf: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/qc8rlhdcdxaf6dwbvv0v4k50w937fyzj-python3-3.10.8/bin/python3"
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
configure flags: --prefix=/nix/store/dc674dk9i612cji6jcqp7xv4gphyyvk1-samba-armv7l-unknown-linux-gnueabihf-4.17.2 --bindir=/nix/store/dc674dk9i612cji6jcqp7xv4gphyyvk1-samba-armv7l-unknown-linux-gnueabihf-4.17.2/bin --sbindir=/nix/store/dc674dk9i612cji6jcqp7xv4gphyyvk1-samba-armv7l-unknown-linux-gnueabihf-4.17.2/sbin --includedir=/nix/store/r8wlb9zbh5p040vdvgrmxwghlwacd8y1-samba-armv7l-unknown-linux-gnueabihf-4.17.2-dev/include --oldincludedir=/nix/store/r8wlb9zbh5p040vdvgrmxwghlwacd8y1-samba-armv7l-unknown-linux-gnueabihf-4.17.2-dev/include --mandir=/nix/store/78dnlml0x4zbwn5a4837igr7pdp7nc5j-samba-armv7l-unknown-linux-gnueabihf-4.17.2-man/share/man --infodir=/nix/store/dc674dk9i612cji6jcqp7xv4gphyyvk1-samba-armv7l-unknown-linux-gnueabihf-4.17.2/share/info --docdir=/nix/store/dc674dk9i612cji6jcqp7xv4gphyyvk1-samba-armv7l-unknown-linux-gnueabihf-4.17.2/share/doc/samba-armv7l-unknown-linux-gnueabihf --libdir=/nix/store/dc674dk9i612cji6jcqp7xv4gphyyvk1-samba-armv7l-unknown-linux-gnueabihf-4.17.2/lib --libexecdir=/nix/store/dc674dk9i612cji6jcqp7xv4gphyyvk1-samba-armv7l-unknown-linux-gnueabihf-4.17.2/libexec --localedir=/nix/store/dc674dk9i612cji6jcqp7xv4gphyyvk1-samba-armv7l-unknown-linux-gnueabihf-4.17.2/share/locale --with-static-modules=NONE --with-shared-modules=ALL --enable-fhs --sysconfdir=/etc --localstatedir=/var --disable-rpath --without-ad-dc --without-ldap --without-ads --with-profiling-data --bundled-libraries=\!asn1_compile\,\!compile_et --jobs 1 --build=x86_64-unknown-linux-gnu --host=armv7l-unknown-linux-gnueabihf
[32m[0mSetting top to                           :[0m [0m[32m[32m/build/samba-4.17.2[0m [0m
[32m[0mSetting out to                           :[0m [0m[32m[32m/build/samba-4.17.2/bin[0m [0m
[32m[0mChecking for 'gcc' (C compiler)          :[0m [0m[32m[32marmv7l-unknown-linux-gnueabihf-gcc[0m [0m
[32m[0mChecking for program 'git'               :[0m [0m[32m[33mnot found[0m [0m
[32m[0mChecking for c flags '-MMD'              :[0m [0m[32m[32myes[0m [0m
[01;31mERROR: Mismatch between --build and --host. Please use --cross-compile instead[0m
```

### Additional context

`samba` cross-compiled correctly with nixpkgs @ 0deec20d46dd26e5fcfc3ce232ea2de94719d75d (~5 weeks ago).

### Notify maintainers

@aneeshusa @ajs124 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
$ nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 5.19.17, NixOS, 22.05 (Quokka), 22.05.20221104.1404483`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - channels(root): `""`
 - channels(majiir): `""`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`
```

**Comment by ajs124** [comment]
I tried switching back to using `wafHook`, but that still fails to build:
```diff
--- a/pkgs/servers/samba/4.x.nix
+++ b/pkgs/servers/samba/4.x.nix
@@ -1,6 +1,7 @@
 { lib, stdenv
 , buildPackages
 , fetchurl
+, wafHook
 , pkg-config
 , bison
 , flex
@@ -63,6 +64,7 @@ stdenv.mkDerivation rec {
 
   nativeBuildInputs = [
     python3Packages.python
+    wafHook
     pkg-config
     bison
     flex
@@ -106,6 +108,8 @@ stdenv.mkDerivation rec {
     ++ optional enableAcl acl
     ++ optional enablePam pam;
 
+  wafPath = "buildtools/bin/waf";
+
   postPatch = ''
     # Removes absolute paths in scripts
     sed -i 's,/sbin/,,g' ctdb/config/functions
@@ -118,9 +122,10 @@ stdenv.mkDerivation rec {
 
   preConfigure = ''
     export PKGCONFIG="$PKG_CONFIG"
+    export PYTHONHASHSEED=1
   '';
 
-  configureFlags = [
+  wafConfigureFlags = [
     "--with-static-modules=NONE"
     "--with-shared-modules=ALL"
     "--enable-fhs"
```

```
/build/samba-4.15.2/bin/compile_et: line 1: ELF��M@�C@8: not found
/build/samba-4.15.2/bin/compile_et: line 2: syntax error: unexpected ")"

/build/samba-4.15.2/bin/compile_et: line 1: ELF��M@�C@8: not found
/build/samba-4.15.2/bin/compile_et: line 2: syntax error: unexpected ")"

/build/samba-4.15.2/bin/compile_et: line 1: ELF��M@�C@8: not found
/build/samba-4.15.2/bin/compile_et: line 2: syntax error: unexpected ")"

/build/samba-4.15.2/bin/compile_et: line 1: ELF��M@�C@8: not found
/build/samba-4.15.2/bin/compile_et: line 2: syntax error: unexpected ")"

/build/samba-4.15.2/bin/compile_et: line 1: ELF��M@�C@8: not found
/build/samba-4.15.2/bin/compile_et: line 2: syntax error: unexpected ")"

/build/samba-4.15.2/bin/compile_et: line 1: ELF��M@�C@8: not found
/build/samba-4.15.2/bin/compile_et: line 2: syntax error: unexpected ")"

/build/samba-4.15.2/bin/compile_et: line 1: ELF��M@�C@8: not found
/build/samba-4.15.2/bin/compile_et: line 2: syntax error: unexpected ")"

/build/samba-4.15.2/bin/compile_et: line 1: ELF��M@�C@8: not found
/build/samba-4.15.2/bin/compile_et: line 2: syntax error: unexpected ")"

/build/samba-4.15.2/bin/compile_et: line 1: ELF��M@�C@8: not found
/build/samba-4.15.2/bin/compile_et: line 2: syntax error: unexpected ")"

/build/samba-4.15.2/bin/compile_et: line 1: ELF��M@�C@8: not found
/build/samba-4.15.2/bin/compile_et: line 2: syntax error: unexpected ")"

Waf: Leaving directory `/build/samba-4.15.2/bin/default'
Build failed
 -> task in 'HEIMDAL_HDB_ERR_ET' failed with exit status 1 (run with -v to display more information)
 -> task in 'HEIMNTLM_ET' failed with exit status 1 (run with -v to display more information)
 -> task in 'HEIMDAL_KRB5_ERR_ET' failed with exit status 1 (run with -v to display more information)
 -> task in 'HEIMDAL_GKRB5_ERR_ET' failed with exit status 1 (run with -v to display more information)
 -> task in 'HEIMDAL_KRB_ERR_ET' failed with exit status 1 (run with -v to display more information)
 -> task in 'HEIMDAL_K524_ERR_ET' failed with exit status 1 (run with -v to display more information)
 -> task in 'HEIMDAL_HEIM_ERR_ET' failed with exit status 1 (run with -v to display more information)
 -> task in 'HEIMDAL_ASN1_ERR_ET' failed with exit status 1 (run with -v to display more information)
 -> task in 'WIND_ERR_ET' failed with exit status 1 (run with -v to display more information)
 -> task in 'HEIMDAL_HX509_ERR_ET' failed with exit status 1 (run with -v to display more information)
```

**Comment by Majiir** [comment]
I ran a bisect, one for `armv7l-hf-multiplatform` and again for `aarch64-multiplatform`. And I rechecked some things manually, because the result makes no sense! Both bisects came up with f7df1e77d56ab4bc51ceca396545d00e81bedc48 as the first bad commit. But I think this is a red herring, because [it's `git` that fails to build](https://gist.github.com/Majiir/1249e2f3b91fb33472fdefb5aec630ae), not `samba`. I will see if I can isolate the confounding failures.
